environs/bootstrap/bootstrap.go

```diff
@@ -230,14 +220,64 @@ func Bootstrap(ctx environs.BootstrapContext, environ environs.Environ, args Boo
 		}
 	}
 
-	ctx.Infof("Bootstrapping model %q", cfg.Name())
-	logger.Debugf("model %q supports service/machine networks: %v", cfg.Name(), supportsNetworking)
-	disableNetworkManagement, _ := cfg.DisableNetworkManagement()
-	logger.Debugf("network management by juju enabled: %v", !disableNetworkManagement)
+	imageMetadata, err := bootstrapImageMetadata(environ,
+		bootstrapSeries,
+		bootstrapArch,
+		args.BootstrapImage,
+		&customImageMetadata,
+	)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	// We want to determine a list of valid architectures for which to pick tools and images.
+	// This includes architectures from custom and other available image metadata.
+	architectures := set.NewStrings()
+	if len(customImageMetadata) > 0 {
+		for _, customMetadata := range customImageMetadata {
+			architectures.Add(customMetadata.Arch)
+		}
+	}
+	if len(imageMetadata) > 0 {
+		for _, iMetadata := range imageMetadata {
+			architectures.Add(iMetadata.Arch)
+		}
+	}
+
+	constraintsValidator, err := environ.ConstraintsValidator()
+	if err != nil {
+		return err
+	}
+
+	// IFF there is any image metadata, then and only then do we intersect
+	// the vocabulary with the architectures from the metadata.
+	if len(architectures) != 0 {
+		architectures.Add(arch.HostArch())
+		constraintsValidator.IntersectVocabulary(constraints.Arch, architectures.SortedValues())
+	}
+
+	bootstrapConstraints, err := constraintsValidator.Merge(
+		args.ModelConstraints, args.BootstrapConstraints,
+	)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	if bootstrapConstraints.Arch == nil {
+		hostArch := arch.HostArch()
+		bootstrapConstraints.Arch = &hostArch
+		// We no longer support controllers on i386.
+		// If we are bootstrapping from an i386 client,
+		// we'll look for amd64 tools.
+		if *bootstrapConstraints.Arch == arch.I386 {
+			amd64 := arch.AMD64
+			bootstrapConstraints.Arch = &amd64
+		}
+	}
 
 	var availableTools coretools.List
 	if !args.BuildAgent {
-		availableTools, err = findPackagedTools(environ, args.AgentVersion, &bootstrapArch, bootstrapSeries)
+		availableTools, err = findPackagedTools(environ, args.AgentVersion, bootstrapConstraints.Arch, bootstrapSeries)
```

Comment:
Reverted but also added comments to ensure that noone else "fixes" it :) and that Andrew's concern is documented :)